util/log/clog.go

```diff
@@ -340,6 +371,99 @@ func (t *traceLocation) Set(value string) error {
 	return nil
 }
 
+// LogEntryDecoder reads successive encoded log entries from the input
+// buffer. Each entry is preceded by a single varint describing the
+// next entry's length.
+type LogEntryDecoder struct {
+	in io.Reader
+}
+
+func NewLogEntryDecoder(in io.Reader) *LogEntryDecoder {
+	return &LogEntryDecoder{in: in}
+}
+
+func (lr *LogEntryDecoder) Decode(entry *proto.LogEntry) error {
+	// Read the next log entry.
+	szBuf := make([]byte, 4)
+	n, err := lr.in.Read(szBuf)
+	if err != nil {
+		return err
+	}
+	_, sz := encoding.DecodeUint32(szBuf)
+	buf := make([]byte, sz)
+	n, err = lr.in.Read(buf)
+	if err != nil {
+		return err
+	}
+	if err := gogoproto.Unmarshal(buf[:n], entry); err != nil {
+		return err
+	}
+	return nil
+}
+
+type baseLogEntryReader struct {
+	buf    []byte
+	ld     *LogEntryDecoder
+	format func(entry *proto.LogEntry) []byte
+}
+
+func (hr *baseLogEntryReader) Read(p []byte) (int, error) {
+	var n int
+	for {
+		if len(hr.buf) != 0 {
+			copied := copy(p, hr.buf)
+			hr.buf = hr.buf[copied:]
+			n += copied
+			p = p[copied:]
+			if len(p) == 0 {
+				return n, nil
+			}
+		}
+		entry := &proto.LogEntry{}
+		if err := hr.ld.Decode(entry); err != nil {
+			return n, err
+		}
+		hr.buf = hr.format(entry)
+	}
+}
+
+type termLogEntryReader struct {
```

Comment:
This type (and jsonLogEntryReader) looks unnecessary; all you need is a baseLogEntryReader with the format field set.